### PR TITLE
add ``UrlStr`` and ``urlstr`` types #236

### LINKED ISSUE #236: Exotic type for urls validation
Hi,

In my projects I need to validate URLs. And I think that it's good to have such type in `pydantic` and I can port URLs validation logic from Django project. 

@samuelcolvin what do you think?

### PR DESCRIPTION
**Pull request by Gr1N**
fix #236

Hi, 

I've ported URL validation from marshmallow library, to my mind it's all we need. @samuelcolvin what do you think?

### CODE REVIEW

File: pydantic/types.py [hunk 1]
```diff
@@ -107,6 +121,60 @@ def validate(cls, value):
         return validate_email(value)[1]
 
 
+class UrlStr(str):
+    strip_whitespace = True
```
Comment (samuelcolvin) [inline pydantic/types.py]: Maybe I'm being dumb, but it looks like `stripe_whitespace = False` won't be respected.

File: pydantic/types.py [hunk 2]
```diff
@@ -107,6 +121,60 @@ def validate(cls, value):
...
+    min_length = 1
+    max_length = 2 ** 16
```
Comment (samuelcolvin) [inline pydantic/types.py]: `min_length` and `max_length` don't seem to have any effect.

File: pydantic/types.py [hunk 3]
```diff
@@ -107,6 +121,60 @@ def validate(cls, value):
...
+    default_schemes = {'http', 'https', 'ftp', 'ftps'}
+    schemes: Optional[Set[str]] = None
```
Comment (samuelcolvin) [inline pydantic/types.py]: I don't think we need `default_schemes`, just

```py
schemes = cls.schemes or {'http', 'https', 'ftp', 'ftps'}
```
below

File: pydantic/types.py [hunk 4]
```diff
@@ -107,6 +121,60 @@ def validate(cls, value):
...
+    @classmethod
+    def validate(cls, value: str) -> str:
+        # Check first if the scheme is valid
+        schemes = cls.schemes or cls.default_schemes
+        if '://' in value:
+            scheme = value.split('://')[0].lower()
+            if scheme not in schemes:
+                raise errors.UrlSchemeError(scheme=scheme)
+
+        regex = url_regex_generator(relative=cls.relative, require_tld=cls.require_tld)
```
Comment (samuelcolvin) [inline pydantic/types.py]: can we generate this on `__init__` then we won't need the lru cache?

File: pydantic/types.py (hunk 3 above)
Comment (samuelcolvin) [reply pydantic/types.py]: or just use `{'http', 'https', 'ftp', 'ftps'}` as the default here and below in `urlstr`.

(using a global variable for `DEFAULT_SCHEMAS` so the actual set isn't duplicated)

File: pydantic/utils.py [hunk 5]
```diff
@@ -165,3 +166,27 @@ def validate_field_name(bases: List[Type['BaseModel']], field_name: str) -> None
         if getattr(base, field_name, None):
             raise NameError(f'Field name "{field_name}" shadows a BaseModel attribute; '
                             f'use a different field name with "alias=\'{field_name}\'".')
+
+
+@lru_cache(maxsize=None)
+def url_regex_generator(*, relative: bool, require_tld: bool) -> Pattern:
```
Comment (samuelcolvin) [inline pydantic/utils.py]: I'm assuming this is taken from marshmallow?

I think we should add a docstring attributing it to marshmallow with a link to where in code it's taken from.

File: tests/test_types.py [hunk 6]
```diff
@@ -1258,3 +1258,514 @@ class JsonDetailedModel(BaseModel):
         'msg': 'JSON object must be str, bytes or bytearray',
         'type': 'type_error.json'
     }
+
+
+@pytest.mark.parametrize(
```
Comment (samuelcolvin) [inline tests/test_types.py]: new file i think.

Also, are these taken from marshmallow or similar? If so I think we should attribute the code/cases.

File: pydantic/types.py (hunk 3 above)
Comment (Gr1N) [reply pydantic/types.py]: fixed

File: tests/test_types.py (hunk 6 above)
Comment (Gr1N) [reply tests/test_types.py]: moved to a new file

File: pydantic/types.py (hunk 1 above)
Comment (Gr1N) [reply pydantic/types.py]: Why? I'm using validators `anystr_strip_whitespace` and `anystr_length_validator` and here a simple example that strip is working:

```
>>> from pydantic import BaseModel, UrlStr
>>> class Model(BaseModel):
...     v: UrlStr
...
>>>
>>> Model(v='http://example.com    ').v
'http://example.com'
>>>
```

File: pydantic/types.py (hunk 4 above)
Comment (Gr1N) [reply pydantic/types.py]: to my mind better to use `lru_cache` in this case, it can help us to decrease number of calls if `re.compile`

**Comment by samuelcolvin** [comment]
Awesome, thank you very much.

Maybe we need some more docs to include `urlstr` and document it's arguments?

**Pull request merged**